### PR TITLE
Set pod environment variables to support torchrun

### PR DESCRIPTION
**Pull request by ChenYi015**

### CODE REVIEW

File: pkg/controller.v1/pytorch/env.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+	EnvPetNprocPerNode = "PET_NPROC_PER_NODE"
+	// EnvPetNnodes is the environment variable name for the number of nodes.
+	EnvPetNnodes = "PET_NNODES"
+	// EnvPetNodeRank is the environment variable name for the rank of nodes.
+	EnvPetNodeRank = "PET_NODE_RANK"
+)
+
+// setPodTemplateEnv sets the environment variables for the podTemplateSpec.
+func setPodTemplateEnv(podTemplateSpec *corev1.PodTemplateSpec, job *v1.PyTorchJob, rtype v1.PyTorchReplicaType, index string) error {
+	rank, err := strconv.Atoi(index)
```
Comment (cheyang) [inline pkg/controller.v1/pytorch/env.go:49]: I suggest adding  unit test for  the new function setPodTemplateEnv to  confirm that it correctly sets all necessary environment variables.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/controller.v1/pytorch/env.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+		if !masterPortFound {
+			podTemplateSpec.Spec.Containers[i].Env = append(podTemplateSpec.Spec.Containers[i].Env, corev1.EnvVar{
+				Name:  EnvMasterPort,
+				Value: strconv.Itoa(int(masterPort)),
+			})
+		}
+
+		if !worldSizeFound {
+			worldSize := int(getTotalReplicas(job))
```
Comment (cheyang) [inline pkg/controller.v1/pytorch/env.go]: I think worldsize = nproc * replicas , by default nproc is 1.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/controller.v1/pytorch/env.go (hunk 1 above)
Comment (ChenYi015) [reply pkg/controller.v1/pytorch/env.go:49]: Done.

File: pkg/controller.v1/pytorch/env.go (hunk 2 above)
Comment (ChenYi015) [reply pkg/controller.v1/pytorch/env.go]: Done. Will read `nprocPerNode` from env in master replica spec and set `worldSize` to `nprocPerNode * totalReplicas`.

File: pkg/controller.v1/pytorch/job.go [hunk 3]
```diff
@@ -244,6 +247,42 @@ func getTotalReplicas(job *pyv1.PyTorchJob) int32 {
...
+		if container.Name != pyv1.DefaultContainerName {
+			continue
+		}
+		for _, env := range container.Env {
+			if env.Name != EnvPetNprocPerNode {
+				continue
+			}
+			val, err := strconv.Atoi(env.Value)
+			if err != nil {
+				nprocPerNode = val
```
Comment (cheyang) [inline pkg/controller.v1/pytorch/job.go]: I suggest logging the error to make it's easy to debug.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/controller.v1/pytorch/job.go (hunk 3 above)
Comment (ChenYi015) [reply pkg/controller.v1/pytorch/job.go]: I think we do not need to log the error as we will ignore it explicitly if the value of `PET_NPROC_PER_NODE` cannot be converted to an integer.

File: pkg/controller.v1/pytorch/env.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+				petMasterPortFound = true
+			case EnvPetNnodes:
+				petNnodesFound = true
+			case EnvPetNodeRank:
+				petNodeRankFound = true
+			}
+		}
+
+		if !masterAddrFound {
+			podTemplateSpec.Spec.Containers[i].Env = append(podTemplateSpec.Spec.Containers[i].Env, corev1.EnvVar{
```
Comment (cheyang) [inline pkg/controller.v1/pytorch/env.go]: I suggest using func `setEnvIfNotFound` to check and set environment variables, reducing code duplication. Just like `setEnvIfNotFound(&podTemplateSpec.Spec.Containers[i].Env, EnvMasterAddr, masterAddr)`

Comment (cheyang) [reply pkg/controller.v1/pytorch/env.go]: Maybe we can enhance it in future.

File: pkg/controller.v1/pytorch/env.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
+// Copyright 2021 The Kubeflow Authors
```
Comment (cheyang) [inline pkg/controller.v1/pytorch/env.go]: 2021 -> 2024

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/controller.v1/pytorch/env.go (hunk 5 above)
Comment (ChenYi015) [reply pkg/controller.v1/pytorch/env.go]: Done.

File: pkg/controller.v1/pytorch/env.go (hunk 4 above)
Comment (ChenYi015) [reply pkg/controller.v1/pytorch/env.go]: Done.

**Review by cheyang — APPROVED** [review]
/lgtm
/approve

**Comment by cheyang** [comment]
Good Job！Thanks.

**Pull request merged**